### PR TITLE
[MASIC][chassis] test_bgp_update_timer change to support for multi-asic

### PR DESCRIPTION
**Pull request by wenyiz2021**
Signed-off-by: Wenyi Zhang [wenyizhang@microsoft.com]

### Description of PR

This PR is to make the test support for multi-asic duts.
To achieve that, some cmds in the test need to specify correct namespace, also, bgp neighbors need to create on specified ASIC(s).
Note that the 2 created pseudoswitchs does not need to be on the same asic, they can still announce routes across asics through backend asics.

Summary:
Fixes # (issue)

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [x] Test case(new/improvement)

### Back port request
- [ ] 201911
- [x] 202012
- [ ] 202205

#### What is the motivation for this PR?
To support test_bgp_update_timer test for multi-asic duts

#### How did you do it?
To achieve that, some cmds in the test need to specify correct namespace, also, bgp neighbors need to create on specified ASIC(s).
Note that the 2 created pseudoswitchs does not need to be on the same asic, they can still announce routes across asics through backend asics.

#### How did you verify/test it?
run this test on multi-asic device:
AzDevOps@8bcda3ef7587:/var/src/sonic-mgmt-int/tests$ sudo ./run_tests.sh -i '../ansible/strsvc,../ansible/veos' -n vmsvc1-t1-masic-acs-1 -u -e "--disable_loganalyzer --skip_sanity" -c bgp/test_bgp_update_timer.py
=== Running tests in groups ===
=================================================================================================================== test session starts ===================================================================================================================

collecting 0 items
------------------------------------------------------------------------------------------------------------------- live log collection -------------------------------------------------------------------------------------------------------------------
18:28:59 bgp. L2530 WARNING| [bgp.py] use_2_bytes_asn: True
collected 1 item

bgp/test_bgp_update_timer.py::test_bgp_update_timer PASSED [100%]

#### Any platform specific information?
This won't change any behavior of single-asic dut

run result for single-asic dut:
AzDevOps@8bcda3ef7587:/var/src/sonic-mgmt-int/tests$ sudo ./run_tests.sh -n vms11-t0-s6000-6 -i ../ansible/str,../ansible/veos  -f ../ansible/testbed.csv -u -e "--disable_loganalyzer --skip_sanity" -l info -k debug -c bgp/test_bgp_update_timer.py
================================================================================================================== test session starts ===================================================================================================================

--------------------------------------------------------------------------------------------- generated xml file: /var/src/sonic-mgmt-int/tests/logs/tr.xml ----------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------------------------- live log sessionfinish -----------------------------------------------------------------------------------------------------------------
19:17:22 __init__.pytest_terminal_summary         L0064 INFO   | Can not get Allure report URL. Please check logs
========================== 1 passed in 542.32 seconds ==========================

### CODE REVIEW

File: tests/bgp/test_bgp_update_timer.py [hunk 1]
```diff
@@ -35,18 +35,18 @@
...
         # have it use LINUX_SLL (Linux cooked) instead.
         start_pcap = "tcpdump -y LINUX_SLL -i %s -w %s port 179" % (iface, save_path)
     else:
-        start_pcap = "tcpdump -i %s -w %s port 179" % (iface, save_path)
-    stop_pcap = "pkill -f '%s'" % start_pcap
-    start_pcap = "nohup %s &" % start_pcap
-    duthost.shell(start_pcap)
+        start_pcap = "tcpdump -i %s -w %s port 179" % (iface, save_path)  
+    # for multi-asic dut, add 'ip netns exec asicx' to the beggining of tcpdump cmd 
+    stop_pcap = "sudo pkill -f '%s %s'" % (duthost.asic_instance_from_namespace(ns).ns_arg, start_pcap)
```
Comment (saiarcot895) [inline tests/bgp/test_bgp_update_timer.py:48]: Will this still work in the single-ASIC case? I'm guessing `duthost.asic_instance_from_namespace(ns).ns_arg` evaluates to an empty string in the single-ASIC case, in which case the command will be `sudo pkill -f ' %s'` (with the space before the `%s`).

Comment (wenyiz2021) [reply tests/bgp/test_bgp_update_timer.py:48]: thanks @saiarcot895, good catch!

Comment (wenyiz2021) [reply tests/bgp/test_bgp_update_timer.py]: @saiarcot895 
new change is working on single-asic dut too, attached log in PR description.
thanks for pointing out!

**Review by saiarcot895 — DISMISSED** [review]

**Comment by mannytaheri** [comment]
I have 3 comments:
1. Default timeout_s is set to 30 seconds in wait_tcp_connection(client, server_hostname, listening_port, timeout_s = 30).
However 30 seconds may be not enough for BGP to establish. We need to increase timeout_s to 60 seconds.

Please add timeout_s=60 in wait_tcp_connection(client, server_hostname, listening_port) on line 78 of tests.common.helpers.bgp.py:
if not wait_tcp_connection(self.ptfhost, self.ptfip, self.port, timeout_s=60):
raise RuntimeError("Failed to start BGP neighbor %s" % self.name)

We had to do similar thing in PR #6206:

pt_assert(wait_tcp_connection(localhost, ptfhost.mgmt_ip, BGP_MONITOR_PORT),
pt_assert(wait_tcp_connection(localhost, ptfhost.mgmt_ip, BGP_MONITOR_PORT, timeout_s=60),

2. Sometimes BGP sessions are not established in 30 seconds.  We may need to change sleep from 30 to 60 on line 201
time.sleep(60)


3. n0.stop_session() and n1.stop_session() on lines 287 and 288 of test_bgp_update_timer do not delete the BGP sessions since stop_session() is not multi asics aware. 
These are the commands which do not delete the BGP sessions: redis-cli -n 4 -c DEL 'BGP_NEIGHBOR|20.0.0.1' redis-cli -n 4 -c DEL 'DEVICE_NEIGHBOR_METADATA|pseudoswitch0'

If you modify stop_session in tests/common/helpers/bgp.py it will work for both single and multi asics.

```
def stop_session(self):
    """Stop the BGP session."""
    logging.debug("stop bgp session %s", self.name)
    if not self.is_passive:
-	self.duthost.shell("redis-cli -n 4 -c DEL 'BGP_NEIGHBOR|%s'" % self.ip)
-       self.duthost.shell("redis-cli -n 4 -c DEL 'DEVICE_NEIGHBOR_METADATA|%s'" % self.name)
+       for asichost in self.duthost.asics:
+           asichost.run_sonic_db_cli_cmd("CONFIG_DB del 'BGP_NEIGHBOR|{}'".format(self.ip))
+           asichost.run_sonic_db_cli_cmd("CONFIG_DB del 'DEVICE_NEIGHBOR_METADATA|{}'".format(self.name))
    self.ptfhost.exabgp(name=self.name, state="absent")
```

**Comment by wenyiz2021** [comment]
> I have 3 comments:
> 
> 1. Default timeout_s is set to 30 seconds in wait_tcp_connection(client, server_hostname, listening_port, timeout_s = 30).
>    However 30 seconds may be not enough for BGP to establish. We need to increase timeout_s to 60 seconds.
> 
> Please add timeout_s=60 in wait_tcp_connection(client, server_hostname, listening_port) on line 78 of tests.common.helpers.bgp.py: if not wait_tcp_connection(self.ptfhost, self.ptfip, self.port, timeout_s=60): raise RuntimeError("Failed to start BGP neighbor %s" % self.name)
> 
> We had to do similar thing in PR #6206:
> 
> pt_assert(wait_tcp_connection(localhost, ptfhost.mgmt_ip, BGP_MONITOR_PORT), pt_assert(wait_tcp_connection(localhost, ptfhost.mgmt_ip, BGP_MONITOR_PORT, timeout_s=60),
> 
> 2. Sometimes BGP sessions are not established in 30 seconds.  We may need to change sleep from 30 to 60 on line 201
>    time.sleep(60)
> 3. n0.stop_session() and n1.stop_session() on lines 287 and 288 of test_bgp_update_timer do not delete the BGP sessions since stop_session() is not multi asics aware.
>    These are the commands which do not delete the BGP sessions: redis-cli -n 4 -c DEL 'BGP_NEIGHBOR|20.0.0.1' redis-cli -n 4 -c DEL 'DEVICE_NEIGHBOR_METADATA|pseudoswitch0'
> 
> If you modify stop_session in tests/common/helpers/bgp.py it will work for both single and multi asics.
> 
> ```
> def stop_session(self):
>     """Stop the BGP session."""
>     logging.debug("stop bgp session %s", self.name)
>     if not self.is_passive:
> -	self.duthost.shell("redis-cli -n 4 -c DEL 'BGP_NEIGHBOR|%s'" % self.ip)
> -       self.duthost.shell("redis-cli -n 4 -c DEL 'DEVICE_NEIGHBOR_METADATA|%s'" % self.name)
> +       for asichost in self.duthost.asics:
> +           asichost.run_sonic_db_cli_cmd("CONFIG_DB del 'BGP_NEIGHBOR|{}'".format(self.ip))
> +           asichost.run_sonic_db_cli_cmd("CONFIG_DB del 'DEVICE_NEIGHBOR_METADATA|{}'".format(self.name))
>     self.ptfhost.exabgp(name=self.name, state="absent")
> ```

thanks @mannytaheri, will address these in next commit

**Comment by wenyiz2021** [comment]
> I have 3 comments:
> 
> 1. Default timeout_s is set to 30 seconds in wait_tcp_connection(client, server_hostname, listening_port, timeout_s = 30).
>    However 30 seconds may be not enough for BGP to establish. We need to increase timeout_s to 60 seconds.
> 
> Please add timeout_s=60 in wait_tcp_connection(client, server_hostname, listening_port) on line 78 of tests.common.helpers.bgp.py: if not wait_tcp_connection(self.ptfhost, self.ptfip, self.port, timeout_s=60): raise RuntimeError("Failed to start BGP neighbor %s" % self.name)
> 
> We had to do similar thing in PR #6206:
> 
> pt_assert(wait_tcp_connection(localhost, ptfhost.mgmt_ip, BGP_MONITOR_PORT), pt_assert(wait_tcp_connection(localhost, ptfhost.mgmt_ip, BGP_MONITOR_PORT, timeout_s=60),
> 
> 2. Sometimes BGP sessions are not established in 30 seconds.  We may need to change sleep from 30 to 60 on line 201
>    time.sleep(60)
> 3. n0.stop_session() and n1.stop_session() on lines 287 and 288 of test_bgp_update_timer do not delete the BGP sessions since stop_session() is not multi asics aware.
>    These are the commands which do not delete the BGP sessions: redis-cli -n 4 -c DEL 'BGP_NEIGHBOR|20.0.0.1' redis-cli -n 4 -c DEL 'DEVICE_NEIGHBOR_METADATA|pseudoswitch0'
> 
> If you modify stop_session in tests/common/helpers/bgp.py it will work for both single and multi asics.
> 
> ```
> def stop_session(self):
>     """Stop the BGP session."""
>     logging.debug("stop bgp session %s", self.name)
>     if not self.is_passive:
> -	self.duthost.shell("redis-cli -n 4 -c DEL 'BGP_NEIGHBOR|%s'" % self.ip)
> -       self.duthost.shell("redis-cli -n 4 -c DEL 'DEVICE_NEIGHBOR_METADATA|%s'" % self.name)
> +       for asichost in self.duthost.asics:
> +           asichost.run_sonic_db_cli_cmd("CONFIG_DB del 'BGP_NEIGHBOR|{}'".format(self.ip))
> +           asichost.run_sonic_db_cli_cmd("CONFIG_DB del 'DEVICE_NEIGHBOR_METADATA|{}'".format(self.name))
>     self.ptfhost.exabgp(name=self.name, state="absent")
> ```



> , timeout_s=60

@mannytaheri for point 1 and 2, how did you come up with the magic number? on an mutli-asic dut I didn't see timeout with default value.
thanks for pointing out point 3, will address this in next commit

**Comment by wenyiz2021** [comment]
latest fix still working for single/multi-asic dut:
single-asic:
```sudo ./run_tests.sh -n vms11-t0-s6000-6 -i ../ansible/str,../ansible/veos  -f ../ansible/testbed.csv -u -e "--disable_loganalyzer --skip_sanity" -l info -k debug -c bgp/test_bgp_update_timer.py
--------------------------------------------------------------------------------------------- generated xml file: /var/src/sonic-mgmt-int/tests/logs/tr.xml ----------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------------------------- live log sessionfinish -----------------------------------------------------------------------------------------------------------------
00:07:32 __init__.pytest_terminal_summary         L0064 INFO   | Can not get Allure report URL. Please check logs
========================== 1 passed in 289.95 seconds ==========================```

**Comment by sanmalho-git** [comment]
FYI - point1 is a wait_until - so it would poll. For point2, instead of hard-coding 30 or 60 seconds, we should use similar wait_until mechanism to poll.

**Comment by wenyiz2021** [comment]
@mannytaheri @sanmalho-git 
one question - I'm seeing this unstable behavior that sometimes the bgp_updates length : 86 after line 230, so leads testcase to fail:
```
                if no_update:
>                   pytest.fail(err_msg)
E                   Failed: no bgp update announce route {'prefix': '10.10.100.0/27', 'nexthop': u'20.0.0.1'} from 20.0.0.2 to 20.0.0.3 
```

in a good case the bgp_updates length is 92.
any idea what caused the instability?

**Comment by mannytaheri** [comment]
> @mannytaheri @sanmalho-git one question - I'm seeing this unstable behavior that sometimes the bgp_updates length : 86 after line 230, so leads testcase to fail:
> 
> ```
>                 if no_update:
> >                   pytest.fail(err_msg)
> E                   Failed: no bgp update announce route {'prefix': '10.10.100.0/27', 'nexthop': u'20.0.0.1'} from 20.0.0.2 to 20.0.0.3 
> ```
> 
> in a good case the bgp_updates length is 92. any idea what caused the instability?

I saw the same issue when I ran the test case on a multi asics chassis but not idea what caused it.

**Comment by wenyiz2021** [comment]
> > @mannytaheri @sanmalho-git one question - I'm seeing this unstable behavior that sometimes the bgp_updates length : 86 after line 230, so leads testcase to fail:
> > ```
> >                 if no_update:
> > >                   pytest.fail(err_msg)
> > E                   Failed: no bgp update announce route {'prefix': '10.10.100.0/27', 'nexthop': u'20.0.0.1'} from 20.0.0.2 to 20.0.0.3 
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > in a good case the bgp_updates length is 92. any idea what caused the instability?
> 
> I saw the same issue when I ran the test case on a multi asics chassis but not idea what caused it.

@mannytaheri I just figured out.
the good behavior comes with n0 and n1 on the same asic, if they are on diff asics then bgp udpate packets are dropped and announce failed.
I am thinking this is due to the `def announce_route()` and `def withdraw_route()` is announcing/withdrawing to exabgp and then advertise back to the same asic, the `across asics` was not able to be achieved

**Comment by wenyiz2021** [comment]
@mannytaheri @sanmalho-git please re-review. somehow I could not add you as reviewers.

- Made n0 and n1 created on the same asics, the other callers of _setup_interfaces_t1_or_t2 are single-asic usage -- they only use the first conn returned and ingore the 2nd. so no behavior impact on other parts
- I continued to use time.sleep(60) in test_bgp_update_timer, instead of wait_until() method, reason being when I code it as wait_until like this :
`pt_assert(wait_until(MAX_TIMEOUT, 5, 0, is_bgp_neigh_created, n0, bgp_facts), 'BGP session {} on duthost is not started after {}sec'.format(n0.name, MAX_TIMEOUT))`
it'll take more than 200sec to execute the wait_until method, guess each finding the comparison function during each execution takes time. For runtime efficiency I continued with old method

**Comment by wenyiz2021** [comment]
@saiarcot895 @arlakshm can you please help review/approve again?

latest log for single asic:
```
bgp/test_bgp_update_timer.py::test_bgp_update_timer[str-s6000-on-2] PASSED                                                                                                         [100%]
=============================================================================== 1 passed in 745.99 seconds ===============================================================================
```

multi-asic is consistently passing:
```
bgp/test_bgp_update_timer.py::test_bgp_update_timer[str-masic-acs-1] PASSED                                                                                                     [100%]

------------------------------------------------------------- generated xml file: /var/src/sonic-mgmt-int/tests/logs/tr.xml --------------------------------------------------------------
=============================================================================== 1 passed in 643.46 seconds ===============================================================================
```

**Review by saiarcot895 — APPROVED** [review]

**Comment by wenyiz2021** [comment]
request for 202205 for chassis

**Pull request merged**